### PR TITLE
* Initial inclusion of .net 4.6.1, c#6 and roslyn script compiler

### PR DESCRIPTION
**Pull request by Argalep**

### CODE REVIEW

**Comment by kevin-10** [comment]
@Vorspire im gonna let you handle this one lol.

**Comment by ggobbe** [comment]
Any news regarding this MR? @Vorspire @kevin-10

**Comment by Xeroxxx** [comment]
We removed the usage of #if MONO statements. This won't work as intended.

**Comment by Tasanar** [comment]
@Vorspire is this good to go or should I close it out. It has been a few months now.

**Comment by dmurphy22** [comment]
I have used `DotNetCompilerPlatform` at work and have had zero problems with it. I think we should go ahead with this. The mono switch is a nice touch as I have no experience with it on mono.

File: Server/ScriptCompiler.cs [hunk 1]
```diff
@@ -210,9 +210,13 @@ public static bool CompileCSScripts(bool debug, bool cache, out Assembly assembl
 
 			DeleteFiles("Scripts.CS*.dll");
 
-			using (var provider = new CSharpCodeProvider())
-			{
-				var path = GetUnusedPath("Scripts.CS");
+#if !MONO
```
Comment (dmurphy22) [inline Server/ScriptCompiler.cs]: Has this actually been tested on mono? We don't want to start adding code that can only be compiled by Roslyn to scripts which in turn will break mono servers.

File: ServUO.exe.config [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8" ?>
 <configuration>
   <startup>
-    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.0" />
+    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.1" />
```
Comment (dmurphy22) [inline ServUO.exe.config]: Should we consider 4.7.2 as it is the latest?

**Comment by dmurphy22** [comment]
Merging into SerUO/Roslyn for testing

**Comment by dmurphy22** [comment]
Tested on Ubuntu 18.04 LTS and scripts compile when using roslyn features using `CSharpCodeProvider`. As @ggobbe mentioned, they support c# 6 features in mono using `CSharpCodeProvider`

I am almost ready to merge this into master. I just want to test on ubuntu 16 and ubuntu 14. Then I want to test .net 4.7.2 for viability.

**Comment by dmurphy22** [comment]
Had no issues with compilation on ubuntu 16 and 14. I consider .net 4.6.1 to be stable as far as compilation goes on all targeted platforms. 

Testing 4.7.2 today.

**Comment by Tasanar** [comment]
Is the plan to eventually merge this into the ServUO master?

**Comment by Tasanar** [comment]
Nm scrolled above.

Does it enhance performance in any way? I imagine it will quite the upgrade.

**Comment by dmurphy22** [comment]
Performance should not be affected as the current codebase will not yet implement features found in the later versions of .net. There is the opportunity however to eventually implement these features to improve performance. Networking has seen improvement in .net 4.7 which would be nice to look into.

Compiling scripts.dll is faster using the new compiler. 

The upgrade _should_ be seamless once I merge probably later today.

File: Server/ScriptCompiler.cs (hunk 1 above)
Comment (dmurphy22) [reply Server/ScriptCompiler.cs]: Tested and works.

**Comment by dmurphy22** [comment]
Tests went well on 4.7.2

This is ready for further community testing. Merging into master.

**Pull request merged**